### PR TITLE
repurpose

### PR DESCRIPTION
**Pull request by danielnorberg**
- Monitor manifest file(s)
- Recreate containers in response to manifest changes
- Kill containers in namespace that are not listed in manifest file(s)
- Internalize container keep-alive loop so that killing the container-agent lets you kill the containers.
- Use standard logging lib.
- Use docker client from docker-stress
- Remove infrastructure containers and container network stack sharing
- Remove manifest version check given substantial capability & behavior divergence
- Log errors instead of exiting on configuration validation failures.

### CODE REVIEW

File: container_agent/docker_client.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+from subprocess import Popen, PIPE
+
+
+log = logging.getLogger(__name__)
+
+
+DEFAULT_DOCKER_ENDPOINT = environ.get('DOCKER_HOST', 'unix:///var/run/docker.sock')
+
+try:
+    p = Popen('which docker', stdout=PIPE, stderr=PIPE, shell=True)
```
Comment (philipcristiano) [inline container_agent/docker_client.py]: Trying to find the location of docker should be moved to the **init** for the docker client and be called in the event the client isn't specified. Making syscalls in the root of the module isn't good, any module that imports this will trigger a syscall. Slowing down the init isn't great but it's better than calling it because of an import

File: container_agent/docker_client.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    out, err = p.communicate()
+    if p.returncode:
+        raise Exception()
+    DEFAULT_DOCKER_CLI = out.strip()
+except:
```
Comment (philipcristiano) [inline container_agent/docker_client.py]: You should be specific about what exceptions should be caught, especially since you raise a general exception above, you might just want to set DEFAULT_DOCKER_CLI above and use that as the else of the conditional as well as expected errors

File: container_agent/docker_client.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        self.err = err
+
+    def __str__(self):
+        return 'docker command failed: %s (%d) out=(%s) err=(%s)' % (self.command, self.code, self.out, self.err)
+
+
+class CliDockerClient(object):
+    def __init__(self, docker=None, endpoint=None):
+        super(CliDockerClient, self).__init__()
+        self.docker = docker or DEFAULT_DOCKER_CLI
```
Comment (philipcristiano) [inline container_agent/docker_client.py]: This will trigger on falsey values passed through on docker, while not valid you should at least honor the arguments passed through and check if `docker is None` instead of checking the truthyness

File: container_agent/docker_client.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    def kill(self, container_id):
+        log.debug('kill %s', container_id)
+        self.cli_check('kill', container_id)
+
+    def destroy(self, container_id):
+        log.debug('destroy %s', container_id)
+        self.cli_check('rm', container_id)
+
+    def list_containers(self, needle=''):
+        if not needle:
```
Comment (philipcristiano) [inline container_agent/docker_client.py:301]: This could be written without the `not` and use the needle if one existed instead of `not` first.

File: container_agent/run_containers.py [hunk 5]
```diff
@@ -26,28 +26,28 @@
...
-constitute a group.  For simplicity, a group shares a network namespace (via
---net=container:<name>).  This means that all containers can see each other as
-"localhost", but it also means that the set of ports they use must be unique
-across the group.  We want to revisit this.
-
 Environmental requirements:
-  - Docker 0.11 or higher (for the --net flag)
-  - Docker daemon runs with =r=false (for safer restart behavior)
+  - Docker 0.11 or higher
+  - Docker daemon runs with -r=false (for safer restart behavior)
```
Comment (philipcristiano) [inline container_agent/run_containers.py:31]: Whitespace changes should be in a separate commit

Comment (danielnorberg) [inline container_agent/run_containers.py:31]: it's not whitespace changes

File: container_agent/docker_client.py (hunk 3 above)
Comment (danielnorberg) [inline container_agent/docker_client.py]: done

File: container_agent/run_containers.py (hunk 5 above)
Comment (philipcristiano) [inline container_agent/run_containers.py:31]: apparently I just don't understand when GH does highlighting for character changes vs line changes

File: container_agent/docker_client.py (hunk 2 above)
Comment (danielnorberg) [inline container_agent/docker_client.py]: not sure how you want this to look

Comment (philipcristiano) [inline container_agent/docker_client.py]: Probably a method on the client called the first time a client is created without an argument, something like

```
def _detect_system_cli(klass):
     location = '/usr/bin/docker'
     try:
          p = Popen('which docker', stdout=PIPE, stderr=PIPE, shell=True)
     except (whatever popen can throw):
          logging.info('error detecting docker cli location, using default of {0}'.format(location))
     else:
         out, err = p.communicate()
         if p.returncode == 0:
           location = out.strip()
     klass._system_cli = location
```

**Comment by philipcristiano** [comment]
+1
